test/cipher.go

```diff
@@ -21,25 +24,274 @@ func HashBench(b *testing.B, hash func() hash.Hash) {
 		c := cipher(key)
 		for j := 0; j < 1024; j++ {
-			c.XORKeyStream(data,data)
+			c.XORKeyStream(data, data)
 		}
 	}
 }
 
 // Benchmark a block cipher operating in counter mode.
+/*
+XXX Broken
 func BlockCipherBench(b *testing.B, keylen int,
-			bcipher func([]byte) cipher.Block) {
+	bcipher func([]byte) cipher.Block) {
 	StreamCipherBench(b, keylen, func(key []byte) cipher.Stream {
 		bc := bcipher(key)
-		iv := make([]byte,bc.BlockSize())
-		return cipher.NewCTR(bc,iv)
+		iv := make([]byte, bc.BlockSize())
+		return cipher.NewCTR(bc, iv)
 	})
 }
+*/
+
+// Compares the bits between two arrays returning the fraction
+// of differences. If the two arrays are not of the same length
+// no comparison is made and a -1 is returned.
+func BitDiff(a, b []byte) float64 {
+	if len(a) != len(b) {
+		return -1
+	}
+
+	mask1 := byte(1)
+	mask2 := byte(2)
+	mask3 := byte(4)
+	mask4 := byte(8)
+	mask5 := byte(16)
+	mask6 := byte(32)
+	mask7 := byte(64)
+	mask8 := byte(128)
+
+	count := 0
+	for i := 0; i < len(a); i++ {
+		if (a[i] & mask1) != (b[i] & mask1) {
+			count += 1
+		}
+		if (a[i] & mask2) != (b[i] & mask2) {
+			count += 1
+		}
+		if (a[i] & mask3) != (b[i] & mask3) {
+			count += 1
+		}
+		if (a[i] & mask4) != (b[i] & mask4) {
+			count += 1
+		}
+		if (a[i] & mask5) != (b[i] & mask5) {
+			count += 1
+		}
+		if (a[i] & mask6) != (b[i] & mask6) {
+			count += 1
+		}
+		if (a[i] & mask7) != (b[i] & mask7) {
+			count += 1
+		}
+		if (a[i] & mask8) != (b[i] & mask8) {
+			count += 1
+		}
+	}
+
+	return float64(count) / float64(len(a)*8)
+}
+
+// Tests a Cipher can encrypt and decrypt
+func BCHelloWorldHelper(t *testing.T,
+	newCipher func([]byte, ...interface{}) abstract.Cipher,
+	n int, bitdiff float64) {
+	text := []byte("Hello, World")
+	cryptsize := len(text)
+
+	bc := newCipher(nil)
+	keysize := bc.KeySize()
+
+	nkeys := make([][]byte, n)
+	ncrypts := make([][]byte, n)
+
+	for i := range nkeys {
+		nkeys[i] = make([]byte, keysize)
+		rand.Read(nkeys[i])
+		bc = newCipher(nkeys[i])
+		ncrypts[i] = make([]byte, cryptsize)
+		bc.Message(ncrypts[i], text, ncrypts[i])
+
+		bc = newCipher(nkeys[i])
+		decrypted := make([]byte, len(text))
+		bc.Message(decrypted, ncrypts[i], ncrypts[i])
+		if !bytes.Equal(text, decrypted) {
+			t.Log("Encryption / Decryption failed", i)
+			t.FailNow()
+		}
+	}
+
+	for i := range ncrypts {
+		for j := i + 1; j < len(ncrypts); j++ {
+			if bytes.Equal(ncrypts[i], ncrypts[j]) {
+				t.Log("Different keys result in same encryption")
+				t.FailNow()
+			}
+
+			res := BitDiff(ncrypts[i], ncrypts[j])
+			if res < bitdiff {
+				t.Log("Encryptions not sufficiently different:", res)
+				t.FailNow()
+			}
+		}
+	}
+}
+
+// Tests a Cipher:
+// 1) Encryption / decryption work
+// 2) Encryption / decryption with different key don't work
+// 3) Changing a bit in the ciphertext or mac results in failed mac check
+// 4) Different keys produce sufficiently random output
+func AuthenticateAndEncrypt(t *testing.T,
```

Comment:
I avoided starting anything with a Test name since it seems to be special
to Go's unit test. I suspect the signature would likely make sure that Go
didn't get confused, but someone reading the code might.

On Tue, Feb 24, 2015 at 10:24 AM, Bryan Ford <notifications@github.com>
wrote:

> In test/cipher.go
> <https://github.com/DeDiS/crypto/pull/38#discussion_r25259871>:
>
> > +
> > +			res := BitDiff(ncrypts[i], ncrypts[j])
> > +			if res < bitdiff {
> > +				t.Log("Encryptions not sufficiently different:", res)
> > +				t.FailNow()
> > +			}
> > +		}
> > +	}
> > +}
> > +
> > +// Tests a Cipher:
> > +// 1) Encryption / decryption work
> > +// 2) Encryption / decryption with different key don't work
> > +// 3) Changing a bit in the ciphertext or mac results in failed mac check
> > +// 4) Different keys produce sufficiently random output
> > +func AuthenticateAndEncrypt(t *testing.T,
>
> "TestAuthenticatedEncryption"?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/DeDiS/crypto/pull/38/files#r25259871>.
>